cookbooks/bcpc-hadoop/templates/default/hdp_yarn-site.xml.erb

```diff
@@ -10,97 +10,117 @@
+    </property>
+
+    <property>
+      <name>yarn.resourcemanager.webapp.address</name>
+      <value><%="#{float_host(@rm_hosts.first[:hostname])}:8088" if not @rm_hosts.empty?%></value>
+    </property>
   <% end %>
 
-  <% @rm_hosts.each do |h| %>   
-  <property>
-    <name>yarn.resourcemanager.address.rm<%= "#{node.chef_environment}#{h[:node_number]}" %></name>
-    <value><%=float_host(h[:hostname])%>:<%=node["bcpc"]["hadoop"]["yarn"]["rm_port"]%></value>
-  </property>
+  <% if node.run_list.expand(node.chef_environment).recipes.include?("bcpc-hadoop::datanode") %>
+    <!-- Group nodemanager restart option with nodemanager port specification
+         since a stable port is needed for a successful restart -->
+    <property>
+      <name>yarn.nodemanager.recovery.enabled</name>
+      <value>true</value>
+    </property>
+
+    <property>
+      <name>yarn.nodemanager.address</name>
+      <value><%=float_host(node[:hostname])%>:<%=node["bcpc"]["hadoop"]["yarn"]["nodemanager"]["port"]%></value>
+    </property>
+
+    <property>
+      <name>yarn.nodemanager.hostname</name>
+      <value><%= float_host(node[:hostname]) %></value>
+    </property>
   <% end %>
-
-  <property> 
-    <name>yarn.resourcemanager.recovery.enabled</name> 
-    <value>true</value> 
-  </property> 
- 
-  <property> 
-    <name>yarn.resourcemanager.store.class</name> 
-    <value>org.apache.hadoop.yarn.server.resourcemanager.recovery.ZKRMStateStore</value> 
-  </property> 
- 
-  <property> 
-    <name>yarn.resourcemanager.zk-address</name>
-    <value><%= @zk_hosts.map{ |s| float_host(s[:hostname]) + ":#{node[:bcpc][:hadoop][:zookeeper][:port]}" }.join(",") %></value>
-  </property>
- 
-  <property> 
-    <name>yarn.client.failover-proxy-provider</name> 
-    <value>org.apache.hadoop.yarn.client.ConfiguredRMFailoverProxyProvider</value> 
-  </property>
-  <% else %>
-
-  <property>
-    <name>yarn.resourcemanager.resource-tracker.address</name>
-    <value><%="#{float_host(@rm_hosts.first[:hostname])}:8031" if not @rm_hosts.empty?%></value>
-  </property>
-
   <property>
-    <name>yarn.resourcemanager.address</name>
-    <value><%="#{float_host(@rm_hosts.first[:hostname])}:#{node["bcpc"]["hadoop"]["yarn"]["rm_port"]}" if not @rm_hosts.empty?%></value>
+    <name>yarn.application.classpath</name>
+    <value>/etc/hadoop/conf,/usr/hdp/current/hadoop-client/*,/usr/hdp/current/hadoop-client/lib/*,/usr/hdp/current/hadoop-hdfs-client/*,/usr/hdp/current/hadoop-hdfs-client/lib/*,/usr/hdp/current/hadoop-yarn-client/*,/usr/hdp/current/hadoop-yarn-client/lib/*</value>
   </property>
 
   <property>
-    <name>yarn.resourcemanager.scheduler.address</name>
-    <value><%="#{float_host(@rm_hosts.first[:hostname])}:8030" if not @rm_hosts.empty?%></value>
+    <name>yarn.nodemanager.aux-services</name>
+    <value>mapreduce_shuffle<% if node.run_list.expand(node.chef_environment).recipes.include?("bach_spark::default") %>,spark_shuffle<% end %></value>
```

Comment:
Will it be a better idea to move all the `aux-services` into an `array` in a node attribute and define corresponding classes for each `aux-service` in a service specific attribute? Something like:
````
default[:bcpc][:hadoop][:yarn][:aux-services] = ["mapreduce_shuffle", "spark_shuffle"]
````
Then have class specific attributes:
````
default[:bcpc][:hadoop][:yarn][:aux-services][:mapreduce_shuffle][:class] = "org.apache.hadoop.mapred.ShuffleHandler"
default[:bcpc][:hadoop][:yarn][:aux-services][:spark_shuffle][:class] = "org.apache.spark.network.yarn.YarnShuffleService"

````
Then we can loop and generate configuration and we do not have to inspect the chef run_list.